### PR TITLE
Store config files in logdir

### PR DESCRIPTION
**Pull request by barrh**
Store configuration files (args.compress) in the output directory, under 'configs' sub-directory.

### CODE REVIEW

**Comment by nzmora** [comment]
It's always advisable to write some explanation about PRs: what/why/how.
It makes it easier (and faster) for us to approach the code.
Thanks
Neta

**Comment by barrh** [comment]
The ideal is: results will be easy to reproduce.
Storing results along with input files is a step in the right direction.
Another advantage is that it makes one's experiments much easier to track. (when conducting multiple experiments in parallel)

Therefore, I had copied the input file specified by args.compress to the logdir.
Ideally, all arguments that are required to reproduce an experiment would live in the same structure as the results.

File: apputils/execution_env.py [hunk 1]
```diff
@@ -20,32 +20,37 @@
...
+def log_execution_env_state(app_argv, *config_fns, logdir=None, gitroot='.'):
     """Log information about the execution environment.
 
     It is recommeneded to log this information so it can be used for referencing
     at a later time.
 
     Args:
-        app_args (dict): the command line arguments passed to the application
+        app_argv (dict): the command line arguments passed to the application
+        config_fns: sequence of filenames to store in logdir, if set
```
Comment (nzmora) [inline apputils/execution_env.py]: We use ```_fn``` to function variables, so this is confusing to me and I prefer you change it.  This is also true in the code below.

File: apputils/execution_env.py [hunk 2]
```diff
@@ -81,7 +86,20 @@ def log_git_state():
...
-    logger.debug("Command line: %s", " ".join(app_args))
+    logger.debug("Command line: %s", " ".join(app_argv))
+
+    if logdir is None:
+        return
+    # clone configuration files to output directory
+    configs_dest = os.path.join(logdir, 'configs')
+    with contextlib.suppress(FileExistsError):
+        os.makedirs(configs_dest)
+    for fn in config_fns:
```
Comment (nzmora) [inline apputils/execution_env.py]: ```for file_name in config_files:```

File: apputils/execution_env.py [hunk 3]
```diff
@@ -81,7 +86,20 @@ def log_git_state():
...
+        if os.path.exists(os.path.join(configs_dest, fn)):
+            logger.debug('{} already exists in logdir'.format(
+                os.path.basename(fn) or fn))
+            continue
```
Comment (nzmora) [inline apputils/execution_env.py]: please replace ```continue``` with ```else:```

File: apputils/execution_env.py [hunk 4]
```diff
@@ -20,32 +20,37 @@
...
     import lsb_release
+    HAVE_LSB = True
 except ImportError:
     HAVE_LSB = False
 
 logger = logging.getLogger("app_cfg")
 
 
-def log_execution_env_state(app_args, gitroot='.'):
+def log_execution_env_state(app_argv, *config_fns, logdir=None, gitroot='.'):
```
Comment (nzmora) [inline apputils/execution_env.py]: Why do we need *config_fns to be a list?  Is this future-proofing or do you use it?
More often than not, future-proofing leads to extra code that rots over time.  If you don't really pass a list of file names, then please simplify.

File: examples/classifier_compression/compress_classifier.py [hunk 5]
```diff
@@ -89,16 +89,20 @@ def main():
...
 
     # Log various details about the execution environment.  It is sometimes useful
     # to refer to past experiment executions and this information may be useful.
-    apputils.log_execution_env_state(sys.argv, gitroot=module_path)
+    if args.compress is None:
+        apputils.log_execution_env_state(sys.argv, gitroot=module_path)
+    else:
+        apputils.log_execution_env_state(sys.argv, args.compress,
+            logdir=msglogger.logdir, gitroot=module_path,
+        )
```
Comment (nzmora) [inline examples/classifier_compression/compress_classifier.py]: * please remove the new-line here (i.e. place the closing parenthesis on the previous line).
* in fact, we don't need the ```else``` clause, do we?  You should handle the case of ```args.compress== None``` in ```apputils.log_execution_env_state```, not here.

File: apputils/execution_env.py (hunk 1 above)
Comment (barrh) [reply apputils/execution_env.py]: config_filename (singular) it is.

File: apputils/execution_env.py (hunk 4 above)
Comment (barrh) [reply apputils/execution_env.py]: I see nothing wrong with future proofing.
going over a loop of no elements in python is usually cleaner than checking if argument is none or not.

File: examples/classifier_compression/compress_classifier.py (hunk 5 above)
Comment (barrh) [reply examples/classifier_compression/compress_classifier.py]: It's correct either way, according to Python's styling guidelines.

**Comment by nzmora** [comment]
Are you rebasing the code?  Please don't do that next time because it turns all comments to "Outdated" and I can't see what you changed.

**Comment by barrh** [comment]
@nzmora I f-pushed with the requested changes.
Also edited the description, and retested.

File: apputils/execution_env.py [hunk 6]
```diff
@@ -20,32 +20,37 @@
...
     import lsb_release
+    HAVE_LSB = True
 except ImportError:
     HAVE_LSB = False
 
 logger = logging.getLogger("app_cfg")
 
 
-def log_execution_env_state(app_args, gitroot='.'):
+def log_execution_env_state(app_argv=None, config_filename=None, logdir=None, gitroot='.'):
```
Comment (nzmora) [inline apputils/execution_env.py]: Why assign a default value to ```app_argv```?  It should always be available in my opinion.

Comment (barrh) [reply apputils/execution_env.py]: In the current context, you are correct. Change it if it bothers you.
I figured it's perfectly useful to call "log_execution_env_state()" with no arguments at all, so why not allow it.

**Comment by barrh** [comment]
there's however, a small issue with the description of the function... lemme f-push correction.

**Comment by barrh** [comment]
@nzmora I moved the call to sys.argv into the function.

**Pull request merged**